### PR TITLE
Improve Local Runtime And Action Review

### LINKED ISSUE #35: Reduce GPU Candidate Build Cost And Export Flakiness
## Problem

Pull-request validation builds both GPU targets with Docker's local image exporter and loads each complete image into the runner before verification. The Terra-derived candidate spends several minutes exporting large CUDA and vLLM layers after the Dockerfile has completed, and one otherwise successful build lost its GitHub-hosted runner during that export.

The vLLM environment is also installed after the application source is copied into both Dockerfiles. A routine source change therefore invalidates the heavy dependency layer even when `images/gpu/vllm-requirements.txt`, the platform base, and runtime arguments are unchanged.

## Solution

- Move GPU dependency installation into a cache boundary determined only by the base image, runtime arguments, and hash-locked vLLM requirements.
- Add one shared in-image validation script for TorchCodec and vLLM imports plus the no-model-weight assertion.
- Use a Docker Buildx `cacheonly` test build for pull-request targets so every Dockerfile instruction and in-image assertion runs without exporting the complete image into the runner's Docker store. See Docker's [cache-only exporter](https://docs.docker.com/build/exporters/#cache-only-export).
- Retain registry-by-digest publication, pull, runtime import, no-weight inspection, immutable-tag verification, and channel promotion on `main`; pull-request optimization must not weaken exact published-candidate validation.
- Add contract tests for the cache boundary, validation script, exporter, and unchanged `main` verification path.

## Acceptance Criteria

- [ ] A source-only change reuses the vLLM dependency layer for both GPU targets.
- [ ] Pull requests build both final filesystems and run the shared import and no-weight checks without a full local image export.
- [ ] `main` still validates the exact digest later published under immutable and channel tags.
- [ ] GPU build failures retain actionable logs, and measured pull-request duration and runner storage use improve.

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct and Contributing Guidelines

### PR DESCRIPTION
**Pull request by PSchmiedmayer**
### :recycle: Current Situation & Problem

The first Carina validation exposed avoidable installation and launch friction, including a fixed Slurm partition assumption, incomplete runtime dependency checks, limited progress feedback, and opaque confirmation identifiers. OpenHands also returns pending tool calls as one confirmation set, while the CLI, web interface, and notebook did not present that set-level decision consistently.

This pull request advances #24, #25, and #26 and resolves #35. Persistent deployment identity and restart recovery remain tracked in #22.

### :gear: Release Notes

- Streamline native and Carina installation with staged progress, storage checks, private persistent directories, automatic Micromamba setup, pinned FFmpeg, and verified vLLM and TorchCodec imports.
- Add a revision-pinned local model snapshot catalog and a gateway-owned download and verification path with serialized concurrent access, atomic publication, checksums, provenance, capacity checks, and symlink rejection.
- Discover and validate Slurm GPU partitions without guessing among ambiguous choices, pass only required non-secret environment variables into allocations, stage models into job scratch, require exact model readiness, disable unsupported FlashInfer sampling by default, and report runtime startup progress and actionable failures.
- Present each OpenHands pending action set as one atomic review in the CLI, web interface, and notebook, with all proposed actions visible before a single allow or reject decision.
- Improve the interactive terminal with arrow-key action review, concise transcripts, elapsed-time status, narrow-terminal support, and backward-compatible identifiers for scripted automation.
- Keep native release archives portable across macOS and Linux by excluding extended attributes.
- Stabilize generic and Terra GPU candidate validation with source-independent, target-scoped vLLM dependency caches, non-persistent uv build archives, one shared in-image verifier, and cache-only pull-request builds while retaining exact-digest pull-and-run verification on `main`.
- Prepare release `0.1.1` across the canonical version, all Python packages and runtime constants, web package metadata, lockfiles, release inputs, native artifacts, and active image examples.

### :books: Documentation

The Carina experiment diary has been replaced with a concise `0.1.1` operational runbook in `docs/carina-cli.md` and a shared interaction guide in `docs/using-heartwood.md`. The README now directs a new installation through the state-aware `heartwood` entry point. Platform status, release guidance, Terra examples, security boundaries, audit semantics, testing expectations, architecture, and the acceptance-driven implementation plan match the implemented behavior and linked follow-up issues. Pre-existing GPU-runtime advisories without patched releases are tracked in #34.

### :white_check_mark: Testing

- 484 Python tests pass with 90.25% total coverage.
- Ruff formatting and linting, mypy, actionlint, ShellCheck, yamllint, REUSE, gitleaks, and documentation link checks pass.
- 42 web unit tests pass with 94.82% statement coverage and 80.77% branch coverage.
- Desktop and narrow-notebook Playwright end-to-end tests pass.
- The versioned `0.1.1` native release artifact installs successfully in a network-disabled Linux container through both generic and emulated Carina paths, including repeat bootstrap, FFmpeg and vLLM layout, checksum, storage-input, permission, and unsafe-manifest failure checks.
- Local Dockerfile checks pass for the generic and platform images. Published-artifact validation on a clean Carina allocation remains the explicit acceptance gate in #25.
- Pull requests execute the generic and Terra GPU import and no-weight assertions through dedicated BuildKit stages without exporting complete local images; `main` still verifies the exact registry digests before tag promotion.

Closes #35.

### Code of Conduct & Contributing Guidelines

By creating and submitting this pull request, you agree to follow our [Code of Conduct](https://github.com/SchmiedmayerLab/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SchmiedmayerLab/.github/blob/main/CONTRIBUTING.md):

- [x] I agree to follow the [Code of Conduct](https://github.com/SchmiedmayerLab/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SchmiedmayerLab/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: packages/gateway/src/heartwood/gateway/_model_snapshots.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+class SnapshotDownloader(Protocol):
+    """Callable contract implemented by ``huggingface_hub.snapshot_download``."""
+
+    def __call__(
+        self,
+        *,
+        repo_id: str,
+        revision: str,
+        local_dir: Path,
+    ) -> str: ...
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/gateway/src/heartwood/gateway/_model_snapshots.py:41]: This is intentional: `SnapshotDownloader` is a `typing.Protocol` callable contract, and the ellipsis is its canonical non-executable method body. No runtime statement is expected here.

File: packages/gateway/src/heartwood/gateway/_gateway.py [hunk 2]
```diff
@@ -510,6 +528,21 @@ def download_model_artifact_now(
...
+        cache_dir: Path | None = None,
+    ) -> Path:
+        """Download a reviewed single-file artifact or multi-file snapshot."""
+        destination = self.model_cache_dir if cache_dir is None else cache_dir
+        try:
+            artifact = self.artifact_catalog.artifact(model_id)
+        except ValueError:
+            snapshot = self.snapshot_catalog.snapshot(model_id)
+            return download_model_snapshot(snapshot, cache_dir=destination)
+        return download_artifact(artifact, cache_dir=destination)
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/gateway/src/heartwood/gateway/_gateway.py:549]: Addressed in 04e78c3. The fallback now catches only `ModelArtifactError`, and a regression test verifies that unrelated catalog `ValueError`s remain visible.

File: packages/cli/src/heartwood/cli/__init__.py [hunk 3]
```diff
@@ -369,16 +391,15 @@ def main(argv: Sequence[str] | None = None) -> int:
...
-            return _event_exit_code(events)
+            directive = "/allow" if args.command in {"allow", "approve"} else "/reject"
+            if args.tool_call_id:
+                directive = f"{directive} {shlex.quote(args.tool_call_id)}"
+            result = InteractiveSession(gateway, session_id=args.session_id).submit(directive)
+            if result.message:
+                print(result.message)
+            if result.events:
+                print(_format_transcript(result.events))
+            return 1 if result.failed else 0
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/cli/src/heartwood/cli/__init__.py:409]: No change is needed here. `InteractionResult.failed` already returns true when any emitted event has kind `error.recorded`, in addition to its explicit `error` flag, so the one-shot path preserves the non-zero exit status. `test_action_decision_returns_failure_for_unknown_pending_action` covers the command-level failure behavior.

[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/cli/src/heartwood/cli/__init__.py:409]: Added the exact gateway-emitted `error.recorded` regression case in a8a12fb; the one-shot command returns 1 and renders the gateway error.

File: docs/web-interface.md [hunk 4]
```diff
@@ -39,15 +39,15 @@ The selected connection must have an available credential reference and an allow
...
+Create a session, select **Detect environment**, and submit the three tasks from the [Terra Jupyter Demo](terra-jupyter-demo.md#start-the-web-interface). Review every member of the proposed action set before selecting **Allow all once**. The reference sequence builds the adult target-condition cohort, fits the explicitly training-only age baseline, and prepares the count-floor-controlled aggregate export.
 
 ![Heartwood synthetic reference analysis showing the cohort, baseline, and aggregate-export conversation](assets/web-reference-analysis.png)
 
 This screenshot is produced by the real reference-analysis system test using the production web build, gateway, OpenHands SDK adapter, repository-verified Skills, persisted session store, and deterministic loopback model fixture. It validates orchestration and interface parity, not model quality or live Terra behavior.
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply docs/web-interface.md]: Addressed in 5bd9b29. The caption now states only the demonstrated workflow scope, interface parity, and explicit model-quality and live-Terra limitations; it no longer describes how the screenshot or document was produced.

File: packages/cli/src/heartwood/cli/_launch.py [hunk 5]
```diff
@@ -286,52 +342,254 @@ def _runtime_environment(env: Mapping[str, str]) -> dict[str, str]:
...
                 payload = json.load(response)
                 if response.status == 200 and payload.get("data"):
                     return True
         except (OSError, ValueError):
+            now = time.monotonic()
+            if now >= next_update:
+                print(f"Still starting vLLM ({int(now - started)} seconds elapsed)...")
+                next_update = now + 15
             time.sleep(1)
     return False
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/cli/src/heartwood/cli/_launch.py]: Addressed in 8cd8ec9. Runtime readiness now requires the requested model identifier, and every successful but empty or mismatched catalog response follows the throttled progress path. Tests cover exact matches, mismatches, and empty catalogs.

File: packages/gateway/src/heartwood/gateway/_model_snapshots.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            encoding="utf-8",
+        )
+        _write_manifest(staging)
+        verify_model_snapshot(staging)
+        _verify_source_record(staging, snapshot)
+        staging.replace(destination)
+    except Exception:
+        shutil.rmtree(staging, ignore_errors=True)
+        raise
+    return destination
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/gateway/src/heartwood/gateway/_model_snapshots.py]: Addressed in 8cd8ec9. A per-snapshot FileLock now covers the destination recheck, download, verification, and atomic publication. filelock is declared directly, lock files are private, and the concurrent-caller regression test verifies one download and shared reuse.

File: packages/webui/src/components/ConversationWorkspace.tsx [hunk 7]
```diff
@@ -193,49 +198,71 @@ const ConversationItem = ({ message }: { message: ConversationMessage }) => {
...
+          variant="outline"
+          onClick={() => onDecision("deny", target)}
+        >
+          <Ban size={16} />
+          Reject all
+        </Button>
+      </div>
+    </section>
+  );
+};
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply packages/webui/src/components/ConversationWorkspace.tsx]: The full batch is already resolved. The public Heartwood command contract carries one member identifier as the action-set representative; OpenHandsSdkBackend.resolve_confirmation validates that identifier, emits resolution events for every pending member, and invokes the upstream set-level decision once. Passing every ID would submit duplicate decisions. Multi-action approve and reject tests cover this path, and 5bd9b29 renames and documents the representative in the web component to make the contract explicit.

File: README.md [hunk 8]
```diff
@@ -35,13 +35,13 @@ See [Architecture](design/03-architecture.md), [Security And Compliance](design/
...
 
 See [Platform Support](docs/platform-support.md) for platform-specific evidence and limitations. All of Us, AnVIL, Seven Bridges, Velsera, DNAnexus, and UK Biobank Research Analysis Platform are design targets rather than supported platforms.
 
-Tagged releases publish a verified native bundle for environments where a platform image is not appropriate. The [latest GitHub Release](https://github.com/SchmiedmayerLab/heartwood/releases/latest) contains the installer, checksum manifest, source bundle, generated notes, and attestations. Install the immutable `0.1.0` release with:
+Tagged releases publish a verified native bundle for environments where a platform image is not appropriate. The [latest GitHub Release](https://github.com/SchmiedmayerLab/heartwood/releases/latest) contains the installer, checksum manifest, source bundle, generated notes, and attestations. Install a selected immutable release with:
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply README.md:44]: No change is required. `0.1.0` is the current published generic release, and the example intentionally demonstrates an immutable install rather than an unresolved placeholder. The protected release workflow runs `verify_release_candidate.py`, which requires this README reference to match the release version before publication. The Carina runbook is handled separately because 0.1.0 does not contain the Carina fixes in this PR.
